### DISCUSSION TITLE
process.env in getStatic/ServerSideProps

### DISCUSSION THREAD

**Discussion by wereHamster** [vercel/next.js#30818] [Help]
The [documentation about env variables ](https://nextjs.org/docs/basic-features/environment-variables) makes this note: 

> Note: In order to keep server-only secrets safe, Next.js replaces process.env.* with the correct values at build time. 

However my tests (Next.js v12) when using `process.env.FOO` in getServerSideProps indicates that the env variable is not replaced at build time but pulled from the process env at run time. Is that a bug in Next.js 12 or a mistake in the documentation?

**Comment by timneutkens** [comment]
It's a mistake in the documentation, only `process.env.NEXT_PUBLIC_*` is replaced at build time, I'll let the team know 👍

**Reply by wereHamster** [reply]
Page using getStaticProps:
```
export default function Page() {
  console.log("Page: NEXT_PUBLIC_FOO", process.env.NEXT_PUBLIC_FOO);
  console.log("Page: CONFIG_FOO", process.env.CONFIG_FOO);
  console.log("Page: FOO", process.env.FOO);

  return null;
}

export const getStaticProps = async () => {
  console.log("getStaticProps: NEXT_PUBLIC_FOO", process.env.NEXT_PUBLIC_FOO);
  console.log("getStaticProps: CONFIG_FOO", process.env.CONFIG_FOO);
  console.log("getStaticProps: FOO", process.env.FOO);

  return { props: {} };
};
```

Page using getServerSideProps:
```
export default function Page() {
  console.log("Page: NEXT_PUBLIC_FOO", process.env.NEXT_PUBLIC_FOO);
  console.log("Page: CONFIG_FOO", process.env.CONFIG_FOO);
  console.log("Page: FOO", process.env.FOO);

  return null;
}

export const getServerSideProps = async () => {
  console.log("getServerSideProps: NEXT_PUBLIC_FOO", process.env.NEXT_PUBLIC_FOO);
  console.log("getServerSideProps: CONFIG_FOO", process.env.CONFIG_FOO);
  console.log("getServerSideProps: FOO", process.env.FOO);

  return { props: {} };
};
````

next.config.js:
```
module.exports = {
  env: {
    CONFIG_FOO: process.env.CONFIG_FOO
  }
}
```

## Case 1: Env variables left blank both during build and run time

```
next build
next start
```

getStaticProps Page:
```
Page: NEXT_PUBLIC_FOO undefined
Page: CONFIG_FOO undefined
Page: FOO undefined
```

getServerSideProps Page (Server):
```
getServerSideProps: NEXT_PUBLIC_FOO undefined
getServerSideProps: CONFIG_FOO undefined
getServerSideProps: FOO undefined
Page: NEXT_PUBLIC_FOO undefined
Page: CONFIG_FOO undefined
Page: FOO undefined
```
Client:
```
Page: NEXT_PUBLIC_FOO undefined
Page: CONFIG_FOO undefined
Page: FOO undefined
```


## Case 2: Env variables blank during built time, set during run time

```
next build
NEXT_PUBLIC_FOO=RUNTIME CONFIG_FOO=RUNTIME FOO=RUNTIME next start
```

getStaticProps Page:
```
Page: NEXT_PUBLIC_FOO undefined
Page: CONFIG_FOO undefined
Page: FOO undefined
```

getServerSideProps Page (Server):
```
getServerSideProps: NEXT_PUBLIC_FOO RUNTIME
getServerSideProps: CONFIG_FOO RUNTIME
getServerSideProps: FOO RUNTIME
Page: NEXT_PUBLIC_FOO RUNTIME
Page: CONFIG_FOO RUNTIME
Page: FOO RUNTIME
```
Client:
```
Page: NEXT_PUBLIC_FOO undefined
Page: CONFIG_FOO undefined
Page: FOO undefined
```

## Case 3: Env variables set during build time, blank during run time

```
NEXT_PUBLIC_FOO=BUILDTIME CONFIG_FOO=BUILDTIME FOO=BUILDTIME next build
next start
```

getStaticProps Page:
```
Page: NEXT_PUBLIC_FOO BUILDTIME
Page: CONFIG_FOO BUILDTIME
Page: FOO undefined
```

getServerSideProps Page (Server):
```
getServerSideProps: NEXT_PUBLIC_FOO BUILDTIME
getServerSideProps: CONFIG_FOO BUILDTIME
getServerSideProps: FOO undefined
Page: NEXT_PUBLIC_FOO BUILDTIME
Page: CONFIG_FOO BUILDTIME
Page: FOO undefined
```
Client:
```
Page: NEXT_PUBLIC_FOO BUILDTIME
Page: CONFIG_FOO BUILDTIME
Page: FOO undefined
```

## Case 4: Env variables set during build time and run time

```
NEXT_PUBLIC_FOO=BUILDTIME CONFIG_FOO=BUILDTIME FOO=BUILDTIME next build
NEXT_PUBLIC_FOO=RUNTIME CONFIG_FOO=RUNTIME FOO=RUNTIME next start
```

getStaticProps Page:
```
Page: NEXT_PUBLIC_FOO BUILDTIME
Page: CONFIG_FOO BUILDTIME
Page: FOO undefined
```
getServerSideProps Page (Server):
```
getServerSideProps: NEXT_PUBLIC_FOO BUILDTIME
getServerSideProps: CONFIG_FOO BUILDTIME
getServerSideProps: FOO RUNTIME
Page: NEXT_PUBLIC_FOO BUILDTIME
Page: CONFIG_FOO BUILDTIME
Page: FOO RUNTIME
```
Client:
```
Page: NEXT_PUBLIC_FOO BUILDTIME
Page: CONFIG_FOO BUILDTIME
Page: FOO undefined
```

---

There appear to be three classes of env variables:

 - **Class A**: Those prefixed with NEXT_PUBLIC_
 - **Class B**: Those registered in next.config.js env
 - **Class C**: All others

Observed behaviour:

- Only **Class A** and **Class B** ever appear in client-side code. The NEXT_PUBLIC_ prefix and env setting in next.config.js act as a safeguard against accidentally exposing sensitive environment variables to the public.
- **Class A** and **Class B** env variables are replaced with `undefined` in client-side code when not set during build time.
- For **Class A** and **Class B** env variables, if set at build time and run time, build time takes precedence.
- Using **Class C** variables in a page that uses getServerSideProps is dangerous as it can lead to mismatches during hydration.
- Using **Class A** and **Class B** variables in a page that uses getServerSideProps is dangerous (hydration mismatch) when not set at build time.

Repo: https://github.com/wereHamster/next-env-testing-ground

**Reply by nawok** [reply]
What is the common way to pass the env variable’s value to the client? I use one for the base URL and the links are broken when I refresh the server-side props via `<Link />`. Only when I reload the page completely I get my links fixed.

**Reply by wereHamster** [reply]
See my notes at the very bottom. It explains which env variables are available to the client code.

**Reply by nickgrato** [reply]
this is amazing, someone should copy / paste this into the docs
